### PR TITLE
Document/comment reduction code

### PR DESCRIPTION
**Pull request by gmarkall**
This is intended to address questions arising in #1400.

### CODE REVIEW

File: numba/cuda/reduction.py [hunk 1]
```diff
@@ -193,7 +239,8 @@ def copy_strides(arr, n, stride, tpb):
         while size >= 16:
             # Find the closest size that is power of two
             p2size = 2 ** int(math.log(size, 2))
-            # Plan for p2size
+            # Generate a plan for invoking kernels to reduce the first p2size
+            # elements
             plan = self._plan(p2size, nbtype, init)
```
Comment (pitrou) [inline numba/cuda/reduction.py]: I'm still confused. If you unfold the code below, you'll see only `plan[0]` is used... So the multiple stages in the `plan` list don't seem useful at all?

Comment (sklam) [inline numba/cuda/reduction.py]: The only reason to pre-populate the plans (https://github.com/gmarkall/numba/blob/reduce_comments/numba/cuda/reduction.py#L333-L338) is to avoid the compilation, which triggers a synchronization, during the reduction.

Comment (pitrou) [inline numba/cuda/reduction.py]: Could that be added as a comment?

**Comment by gmarkall** [comment]
I've pushed some more changes but they break the pre-compilation of all required kernels prior to starting the reduction operation - more to follow...

**Comment by gmarkall** [comment]
Thanks for the earlier comments - I had misunderstood what was going on with `_plan` - at first I tried to make the code clearer, but ended up thinking that whilst I'm digging into the code, I might as well try and fix the limitation where it uses the host for finishing the reduction, so I've had a stab at fixing this. The main changes to do this are:

- The reduction template kernels are not compiled as lazily anymore - kernels for all block sizes are compiled upon the first use of a reducer.
- A fix to the shared memory size computation for smaller block sizes, where OOB access was taking place.
- `__call__` now has an extra parameter, `res`, in which you can pass a device array to store the result. This saves any device->host data transfer. If you don't pass in `res`, you still get the result returned (this is so that the use-case where you want to reduce a large host array is still possible without any data transfer code needing to be added to user code).
- Addition of tests of edge-cases for the input sizes and block sizes, testing that the initial value is respected, and testing that storing the result into `res` functions correctly.
- Documentation updates to reflect the changes.

File: numba/cuda/reduction.py [hunk 2]
```diff
@@ -133,37 +163,15 @@ def _type_and_size(self, dary, size):
...
-    def _partial_inplace_driver(self, dary, size, init, stream):
+    def _execute_reduction(self, dary, size, stream, init=0):
 
         from numba import cuda
 
         nbtype, size = self._type_and_size(dary, size)
 
+        # Partial reduction kernels write their output strided - this kernel is
+        # used to copy the strided output to a compacted array which can be used
+        # as input to the next kernel.
```
Comment (pitrou) [inline numba/cuda/reduction.py]: As a side note, do you know why a shared variable (`sm`) is used below instead of a local variable? It doesn't seem to be sharing anything useful, and it requires a syncthreads() call.

Comment (gmarkall) [inline numba/cuda/reduction.py]: I don't see a reason for it - I'll try removing it.
